Design.md

```diff
@@ -136,46 +121,88 @@ This API will update the existing storage tenant with the provided `cd_tenant_id
-* Use the tenant account credentials from the `assumeRoleCache` cache for the given account ID before each User API.
-* If the tenant account credentials were not found in `assumeRoleCache`, `AssumeRoleBackbeat` api needs to be called as Super Admin 
-* `assumeRoleCache` cache will be key-value pair of account_id and temporary credentials object respectively.
-    * `assumeRoleCache` cache will have TTL.
-    * Refresh the credentials periodically on the cache (AWS SDK `refresh()` will be used)
-    * _[ A detailed cache design needs to be added ]_
+### Common Behavior for All User APIs
+* Invoke the [Assume Role flow](#Assume-Role) before accessing any user APIs. 
+* If any user API returns an `Access Denied` error with the `user [RoleArn] don't have any policies, denied access`  message,  then invoke the `SetupAssumeRolePolicy` flow (See [Setup-Assume-Role-Policy](#Set-up-assume-role-policy).
+
+### Set Up Assume Role Policy
+1. Generate an access key for the account by using vaultclient to call `generate-account-access-key` with `expirationTime`.
+1. Invoke IAM `get-policy` API for policy `adminPolicy@[account-id]` using the generated access key.
+1. If no policy is returned, use the generated access key to create an IAM managed policy `adminPolicy@[account-id]` with full s3 and iam access (`{ s3:*, iam:* }`).
+1. Invoke `attach-role-policy` to attach the policy `adminPolicy@[account-id]` to the `osis` role using the generated access key.
+1. `delete-access-key` for the account.
 
+### Assume Role
+* Before invoking a User API, you must call the `AssumeRole` flow. Use the tenant account credentials from `assumeRoleCache` for the given account ID before each User API.
+* If the tenant account credentials were not found in `assumeRoleCache`, the `AssumeRoleBackbeat` API must be called as superadmin and added to the `assumeRoleCache` cache.
+* If the `AssumeRoleBackbeat` API returns a `Role does not exist` error, then invoke the [`SetupAssumeRole` subroutine](#SetupAssumeRole-subroutine)).
+* `assumeRoleCache` cache will be a **[key, value]** pair of **[Role_Arn, Temporary_Credentials]** respectively.
+    * Each `assumeRoleCache` entry will have a 50-minute TTL (its session token is valid only for 60 minutes).
+    * When accessing credentials on the cache, refresh the credentials as well as `needsRefresh` (using AWS SDK `refresh()`), as they are subject to expiration.
+    * Cache design can be found [here](#Cache-Design).
+    
 ### Create User
 This API creates a user on Vault.
 1. `create-user` api will be called using assumed role credentials.
-    * `cd_user_id` is stored as username.
+    * The tenant user's `cdUserId` is stored as `username` in the corresponding Vault user.
+    * The tenant user's `username` and `role` enum value are stored in the Vault user path as `<tenantUsername>::<roleEnumValue>`.
 1. `attach-policy` for the new user will be called using the assumed role credentials with full S3 and IAM access permissions.
```

Comment:
Is this still true?

Reply:
This is an attach policy to the user so user can login on OSIS and do S3 operations. IAM is not required, I will remove it. Thanks

Reply:
done 569dba03652832856a3dc38329c6cbf4872fa1d4